### PR TITLE
sc3-plugins: init at 3.11.1

### PR DESCRIPTION
**Pull request by JeffreyBenjaminBrown**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lilyinstarlight** [comment]
Hi there! I've actually had a package of sc3-plugins for a while that I've been using in https://github.com/lilyinstarlight/foosteros since newer versions of Sonic Pi require it.

One thing in particular I noticed was that SuperCollider will not load those plugins by default (and Sonic Pi therefore won't use them) since they are not in the SuperCollider directory in the Nix store, so I actually made a patch for SuperCollider to allow a plugin system to work like this which makes a derivation that automatically loads the plugins:
```nix
with pkgs;

supercollider.override {
  plugins = with supercolliderPlugins; [ sc3-plugins ];
}
```

I just pushed the changes I had locally (that I'd been sitting on for a while) so you can see what I did: https://github.com/NixOS/nixpkgs/compare/master...lilyinstarlight:feature/supercollider-plugins

Do the plugins automatically load for you without making such a change?

If not, would you be interested in adding changes to this PR that allow the plugins to be automatically loaded for programs such as Sonic Pi?

**Comment by JeffreyBenjaminBrown** [comment]
That sounds good to me, Lily! The plugins don't load for me by default, so I've been pfaffing around with symlinks and the `ugenPluginsPath` server option (`-U` at the command line) instead of having SuperCollider load from the system folder like it does by default. 

I'm not sure how to reconcile what looks like a pull request you've made to `nixpkgs` with the one I made. You're suggesting a way to modify the one I made, right? Yours looks cooler -- it's got tests! -- and I'm not looking for authorship. If your pull request is in fact better I'm happy to cede mine for yours.

**Comment by lilyinstarlight** [comment]
Yeah I initially tried injecting the `-U` command line option, but things were still not loading right or not loading everything (I mostly use downstream applications that use SuperCollider so I'm not really a SuperCollider expert). I just supplied a patch allowing pulling the system directories from environment variables at runtime instead of requiring they be hard-coded into the executable to make it load like it would on other operating systems if sc3-plugins were installed

What I linked was intended to be a pull request but you uploaded yours before I finally got back to fixing up my changes and making them good enough for a pull request. I had tests I was using in the package in https://github.com/lilyinstarlight/foosteros that I ported over yesterday which confirm that at least `MdaPiano` loads from the supercollider-with-plugins derivation

If you are good with it and have no further input on my changes that add sc3-plugins, I'm happy to make a new PR for them!

**Comment by lilyinstarlight** [comment]
I've done a touch more clean-up and opened #147011 with the plugin updates. I would love any feedback you might have @JeffreyBenjaminBrown there and would also love if you could test it out and make sure it works for you!

**Comment by JeffreyBenjaminBrown** [comment]
Using your pull request instead of mine sounds fine to me :) My only concern is that it might take longer to be incorporated into nixpkgs? I don't really know how the review process works but when I first submitted this request a month ago there were a lot of red Xs somewhere and now they've turned to green checkmarks (or something like that -- I only remember the semantics, not the actual symbols). But if you don't think that's an issue, sure, take over!

I'll try to test your derivation soon.

**Comment by lilyinstarlight** [comment]
The review process is documented at https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions and while I'm not the most familiar with nixpkgs workflows, having other people test out new packages and providing a review helps speed things along

My PR will be waiting on the `supercollider` package maintainer (or someone with similar authority) to okay the changes to the existing package so it could take longer, but I'm not too worried if the sc3-plugins package gets tested by others. The automated checks have already completed and ran the tests I added as well, so that should help too

Thank you for making this PR, though! (Also thanks for providing the motivation to clean up my changes for PR)

**Comment by agrathwohl** [comment]
Hello @JeffreyBenjaminBrown, I have been lurking around this PR since you first submitted it, and am very appreciative of your efforts to bring this essential SC3 package to NixOS!

After taking a look at your `default.nix` for sc3-plugins again recently, I noticed that you [disable support for Supernova](https://github.com/JeffreyBenjaminBrown/nixpkgs/blob/ea717c7b8f9ff2fbc1e10e73684d25ab1850bd06/pkgs/development/libraries/sc3-plugins/default.nix#L38) in the cmake flags. Is there an explicit need to do this to make the package work correct? I ask because the `supercollider` package on NixOS works great with Supernova, which I use almost exclusively on SC3 now. Might there be a way to enable/disable supernova support?

Though I am new to the NixOS world, I would be happy to take a stab at contributing a solution, but first wanted to check in and make sure this wouldn't counteract your efforts in any way. Thanks again for doing the hard work that I failed to do successfully!

**Comment by JeffreyBenjaminBrown** [comment]
@agrathwohl I don't know why I did that -- probably monkey-pasted someone else's code -- and it appears to be unnecessary. @lilyinstarlight's [recipe](https://github.com/NixOS/nixpkgs/pull/147011), which almost surely* puts mine to shame, [enables Supernova](https://github.com/NixOS/nixpkgs/blob/70e93bc3875e38b07324437a94f241c1137b55e4/pkgs/development/interpreters/supercollider/plugins/sc3-plugins.nix).

*I'm afraid I still haven't tested it like I promised.

**Comment by lilyinstarlight** [comment]
> I don't know why I did that. I probably monkey-pasted someone else's code. It appears to be unnecessary, because @lilyinstarlight's recipe, which almost surely* puts mine to shame, enables Supernova.

Monkey-pasting code can be a part of learning how to use a new language or system 🙂

I do feel bad for kinda swooping in with a different set of proposed changes after you opened this one, but there's no shame in your changes. If we need to split the sc3-plugins addition from the supercollider changes (keeping the sc3-plugins stuff in this PR), I'm still happy to work with you on that too

> *I'm afraid I still haven't tested it like I promised

Whenever you are able to is fine! It would probably help it get merged quicker, but it is still waiting on someone with authority (maintainer or otherwise) to comment on the proposed supercollider changes

**Comment by JeffreyBenjaminBrown** [comment]
> Monkey-pasting code can be a part of learning how to use a new language or system

I believe that. Particularly when you've read a ton of documentation, still don't feel like you're missing a lot of fundamentals, and the returns to reading more appear to have diminished but might rise in the future as more documentation appears.

> I do feel bad for kinda swooping in

No no no, swoop away, crush it. I just want the thing to exist; I don't care whose it is.

File: pkgs/development/libraries/sc3-plugins/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  version = "3.11.1";
+
+  meta = with lib; {
+    description = "Synths and other tools for SuperCollider";
+    homepage = "https://supercollider.github.io/";
+    license = "GPL version 2.0";
+    branch = "Version-3.11.1";
+    platforms = lib.platforms.linux;
+    maintainers = [ maintainers.jeffBrown ];
+  };
```
Comment (urandom2) [inline pkgs/development/libraries/sc3-plugins/default.nix:20]: generally `meta` is the last section in the derivation

File: pkgs/development/libraries/sc3-plugins/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Based on
+# https://gist.github.com/gosub/a42e265ec38d9df203d6
+
+{ stdenv, lib, fetchgit,
+  cmake,
+  supercollider, fftw, libsndfile }:
+
+stdenv.mkDerivation {
+
+  name = "sc3-plugins";
```
Comment (urandom2) [inline pkgs/development/libraries/sc3-plugins/default.nix:10]:
```suggestion
  pname = "sc3-plugins";
```

File: pkgs/development/libraries/sc3-plugins/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  src = fetchgit {
+    url = "https://github.com/supercollider/sc3-plugins";
+    rev = "209cf4ffdcc9181b37aedbf4902e4b4d4090d505"; # This is an ordinary git commit hash, the one corresponding to tag "Version-3.11.1".
+    sha256 = "1cy7g2mvmikml4dg6v4fzw6qr2yv9c94531iwxp501fr9j6z5jh8";
+      # To determine this sha256 value, I first supplied a garbage one
+      # and ran `nix-build -L` in the folder containing this file.
+      # The resulting error suggested this one.
+    fetchSubmodules = true;
+  };
```
Comment (urandom2) [inline pkgs/development/libraries/sc3-plugins/default.nix:30]: is there a reason we are using `fetchgit` not `fetchFromGitHub`?
```suggestion
  src = fetchFromGItHub {
    owner = "supercollider";
    repo = pname;
    rev = "Version-${version}";
    hash = "sha256-CMryjUzZBVBu5zGMQhJL24uMDf+ObPMaoXXGuqt4x7M=";
    fetchSubmodules = true;
  };
```

File: pkgs/development/libraries/sc3-plugins/default.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  buildInputs = [ cmake
+                  supercollider
+                  fftw
+                  libsndfile ];
```
Comment (urandom2) [inline pkgs/development/libraries/sc3-plugins/default.nix:35]: use `nixpkgs-fmt` syntax:
```suggestion
  buildInputs = [
    cmake
    supercollider
    fftw
    libsndfile
  ];
```

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -5230,6 +5230,12 @@
     githubId = 9425955;
     name = "Jean-François Labonté";
   };
+  jeffBrown = {
```
Comment (urandom2) [inline maintainers/maintainer-list.nix:5233]: it is preferred that maintainer changes happen as a separate pr or commit, your call

**Comment by JeffreyBenjaminBrown** [comment]
Lily's competing package mentioned above blows this one out of the water. I should have closed this pull request once hers worked. Now I have. Thanks and sorry for the time you spent on it, @urandom2

**Pull request closed without merging**